### DISCUSSION TITLE
Hi

### DISCUSSION THREAD

**Discussion by prof-cocoon** [marcel-licence/ml_synth_sampler_example#2] [Q&A]
Hi Marcel, finally i've managed to get this project working on a wemos lolin d32 pro!
This message is to talk about a few comments/troubles i've come across...some of them i've found the solution by myself but some persist and i've no idea to solve them.
- it seems there are a lot of serial messages during load of a soundfont, and they are very large in complex and large soundfonts, making the load very slow (i think this is the reason, because large but simple sf2 are loaded in seconds from a regular sd card). Is there a way of not send so much messages? is there a way of providing % of loading progress to display? i'm asking this because loading times might be annoying but if they are the minimum possible it's ok to wait, but without reading the serial output i've no idea if the loading is stuck or it's getting a lot of time (i'm using a touch-tft display to control the file operations).

- what about polyphony? it seems to be 10 voices the max. I've disabled all effect processing code in order to get the sampler player as simple as possible, but i don't know if that will allow more voices to be processed in time. Also i have not found anywhere in the project nor ML Tools where this limit of 10 voices is set.

- what about multi-timbrality? the program seems to listen to every midi channel. is there a way to assign different files to specific channels? I know this can be pre-set from the sf2 files previously, but it would be great if it would be more flexibility in channel assignment. 

- what about adsr? I see that hold and release data is passed from sf2 info to the sampler, but no clues about attack, decay and sustain. they are extremely important to proper sound shaping.

- what about others modifiers as vibrato or tremolo? again, they are a key factor of many sf2, from strings to organs.

thank you a lot for your hard work, greetings from Argentina, Nicolas.

(this is a pic of the project going..)
![image](https://github.com/marcel-licence/ml_synth_sampler_example/assets/88905855/af39134a-733a-420f-97fd-385dc550d06a)

**Comment by marcel-licence** [comment]
Hi Nicolas, 
I am happy to hear that the project is working for you.

- I could make the debugoutput switchable or assignable to a custom function to allow to generate messages on a display etc. That would be a new feature.
- the polyphony is covered by the sampler. I thought it should be limited to 32. I will check where the limit comes from you've observed
- multi-timbral: the sampler can play all samples stored in RAM. There is no difference using the same instrument or different instruments. If all instruments are stored in RAM, they can be played. I had a version running allowing program change messages to access the different instruments pre-loaded. Each channel can use another instrument. It should be very flexible.
- Implementing the complete ADSR might reduce the max poly a lot. I think the release should be implemented and quite matching (I've tested it some weeks ago). 
- vibrato and tremolo is intentionally missing. I wanted to get access to soundfonts / samples without recreating a complete soundfont core. I am not sure how good it would work when trying to implement the full-blown support of all features. It may be possible to add some post effects to get into the direction

I could imagine that some update can be made =)

Best wishes (from Germany),
Marcel
